src/main/java/org/ethereum/beacon/discovery/schema/NodeRecord.java

```diff
@@ -47,36 +47,44 @@ public class NodeRecord {
   private final IdentitySchemaInterpreter identitySchemaInterpreter;
 
   private NodeRecord(
-      IdentitySchemaInterpreter identitySchemaInterpreter, UInt64 seq, Bytes signature) {
+      IdentitySchemaInterpreter identitySchemaInterpreter,
+      UInt64 seq,
+      Bytes signature,
+      Map<String, Object> fields) {
     this.seq = seq;
     this.signature = signature;
     this.identitySchemaInterpreter = identitySchemaInterpreter;
-  }
-
-  private NodeRecord(IdentitySchemaInterpreter identitySchemaInterpreter, UInt64 seq) {
-    this.seq = seq;
-    this.signature = MutableBytes.create(96);
-    this.identitySchemaInterpreter = identitySchemaInterpreter;
+    this.fields.putAll(fields);
+    // serialise to check size
+    Bytes serializedNodeRecord = serialize();
```

Comment:
Wouldn't it be nice if the NodeRecord was immutable, then we could memoize the serialized record.
It is good that we are checking the resulting length here, but there is not guarantee that the set() method is not called, or the setSignature().
I haven't checked whether we are using the set method, but it would be good to get rid of it!
And what is the business with the signature?
I guess if we don't make the NodeRecord immutable we should check the length when the serialize is called.